discovery/consul/consul.go

```diff
@@ -203,21 +235,49 @@ func (d *Discovery) shouldWatch(name string) bool {
 	return false
 }
 
+// shouldWatch returns whether the service of the given name should be watched based on its tags.
+// This gets called when the user doesn't specify a list of services in order to avoid watching
+// *all* services. Details in https://github.com/prometheus/prometheus/pull/3814
+func (d *Discovery) shouldWatchFromTags(tags []string) bool {
+	// If there's no fixed set of watched tags, we watch everything.
+	if d.watchedTag == "" {
+		return true
+	}
+
+	for _, tag := range tags {
+		if d.watchedTag == tag {
+			return true
+		}
+	}
+	return false
+}
+
+// Get the local datacenter if not specified.
+func (d *Discovery) getDatacenter() error {
+	// If the datacenter was not set from clientConf, let's get it from the local Consul agent
+	// (Consul default is to use local node's datacenter if one isn't given for a query).
+	if d.clientDatacenter != "" {
+		return nil
+	}
+
+	info, err := d.client.Agent().Self()
+	if err != nil {
+		level.Error(d.logger).Log("msg", "Error retrieving datacenter name", "err", err)
+		rpcFailuresCount.Inc()
+		return err
+	}
+
+	d.clientDatacenter = info["Config"]["Datacenter"].(string)
```

Comment:
This will panic if this is not a string, you should check the type here.

Reply:
Done